horizomer/utils/tree.py

```diff
@@ -13,6 +13,119 @@
 #
 
 from skbio import TreeNode
+from math import isclose
+from skbio.tree import MissingNodeError
+
+
+def _compare_length(node1, node2):
+    """Private function for compare_branch_lengths. Determines if lengths of the
+    two nodes are same.
+
+    Parameters
+    ----------
+    node1: skbio.TreeNode
+        first node to compare
+    node2: skbio.TreeNode
+        second node to compare
+
+    Returns
+    -------
+    bool
+        `True` if lengths of the two input nodes are None, same, or close
+        `False` otherwise
+
+    See Also
+    --------
+    compare_branch_length
+
+    Examples
+    --------
+    >>> from skbio import TreeNode
+    >>> tree = TreeNode.read(["(a:1,b:1)c;"])
+    >>> print(_compare_length(tree.find('a'), tree.find('b')))
+    True
+    >>> print(_compare_length(tree.find('a'), tree.find('c')))
+    False
+
+    """
+    if node1.length is None and node2.length is None:
+        return True
+    elif (node1.length is None) ^ (node2.length is None):
+        return False
+    elif isclose(node1.length, node2.length) is False:
+        return False
+    return True
+
+
+def compare_branch_lengths(tree1, tree2):
+    """Returns `True` if each corresponding node in 2 trees has same length.
+
+    Parameters
+    ----------
+    tree1: skbio.TreeNode
+        first tree to compare
+    tree2: skbio.TreeNode
+        second tree to compare
+
+    Returns
+    -------
+    bool
+        `True` if two input trees have same topologies and branch lengths
+        `False` otherwise
+
+    See Also
+    --------
+    compare_topology
+
+    Examples
+    --------
+    >>> from skbio import TreeNode
+    >>> tree1 = TreeNode.read(['((a:1, c:1):2, b:1);'])
+    >>> tree2 = TreeNode.read(['((a:1, c:1):2, b:1);'])
+    >>> print(compare_branch_lengths(tree1, tree2))
+    True
+    >>> tree3 = TreeNode.read(['((a:1, c:1):2, b);'])
+    >>> print(compare_branch_lengths(tree3, tree1))
+    False
+    """
+    tcopy1 = tree1.copy()
+    tcopy2 = tree2.copy()
+    stack = []  # stack to store nodes in tree2
+    count = 0
+
+    for node in tcopy1.postorder(include_self=False):
+        if node.is_tip():
+            try:
+                cur = tcopy2.find(node.name)
+            except MissingNodeError:
+                return False
+            if _compare_length(node, cur) is False:
+                return False
+            if node.parent.id is None and cur.parent.id is None:
```

Comment:
the next few lines seem to be a code duplication between the cases of being a tip or an internal node. Is it possible to move them up and thus merge?